cmd/consumer/main.go

```diff
@@ -0,0 +1,175 @@
+}
+
+func New() *RecordProcessor {
+	return &RecordProcessor{
+		sleepDuration:     5 * time.Second,
+		checkpointRetries: 5,
+		checkpointFreq:    60 * time.Second,
+	}
+}
+
+func (rp *RecordProcessor) Initialize(shardID string) error {
+	rp.shardID = shardID
+	rp.lastCheckpoint = time.Now()
+	return nil
+}
+
+func (rp *RecordProcessor) checkpoint(checkpointer kcl.Checkpointer, sequenceNumber string, subSequenceNumber int) {
+	for n := 0; n < rp.checkpointRetries; n++ {
+		if err := checkpointer.Checkpoint(sequenceNumber, subSequenceNumber); err == nil {
+			break
+		} else {
+			if cperr, ok := err.(kcl.CheckpointError); ok {
+				switch cperr.Error() {
+				case "ShutdownException":
+					fmt.Fprintf(os.Stderr, "Encountered shutdown exception, skipping checkpoint\n")
+					return
+				case "ThrottlingException":
+					if rp.checkpointRetries-1 == n {
+						fmt.Fprintf(os.Stderr, "Failed to checkpoint after %d attempts, giving up.\n", n)
+						return
+					}
+					fmt.Fprintf(os.Stderr, "Was throttled while checkpointing, will attempt again in %s", rp.sleepDuration)
+				case "InvalidStateException":
+					fmt.Fprintf(os.Stderr, "MultiLangDaemon reported an invalid state while checkpointing\n")
+				}
+			}
+			fmt.Fprintf(os.Stderr, "Encountered an error while checkpointing: %s", err)
+		}
+		time.Sleep(rp.sleepDuration)
+	}
+}
+
+func (rp *RecordProcessor) shouldUpdateSequence(sequenceNumber *big.Int, subSequenceNumber int) bool {
+	return rp.largestSeq == nil || sequenceNumber.Cmp(rp.largestSeq) == 1 ||
+		(sequenceNumber.Cmp(rp.largestSeq) == 0 && subSequenceNumber > rp.largestSubSeq)
+}
+
+func (rp *RecordProcessor) ProcessRecords(records []kcl.Record, checkpointer kcl.Checkpointer) error {
+	for _, record := range records {
+		// Base64 decode the record
+		data, err := base64.StdEncoding.DecodeString(record.Data)
+		if err != nil {
+			return err
+		}
+		msg := string(data)
+
+		// TODO: Add additional "decoding"
+		// - pull JSON data out of string
+		// - parse RSyslog format
+		// - etc...
+
+		// Write the message to firehose
+		err = rp.firehoseWriter.ProcessMessage(msg)
+		if err != nil {
+			return err
+		}
+
+		// Handle checkpointing
+		// TODO: How to handle the difference between ProcessMessage (sent to FirehoseOutput) vs successfully sent?
+		seqNumber := new(big.Int)
+		if _, ok := seqNumber.SetString(record.SequenceNumber, 10); !ok {
+			fmt.Fprintf(os.Stderr, "could not parse sequence number '%s'\n", record.SequenceNumber)
+			continue
+		}
+		if rp.shouldUpdateSequence(seqNumber, record.SubSequenceNumber) {
+			rp.largestSeq = seqNumber
+			rp.largestSubSeq = record.SubSequenceNumber
+		}
+	}
+	if time.Now().Sub(rp.lastCheckpoint) > rp.checkpointFreq {
+		rp.checkpoint(checkpointer, rp.largestSeq.String(), rp.largestSubSeq)
+		rp.lastCheckpoint = time.Now()
+
+		// Write status to file
+		err := appendToFile("/tmp/kcl_stderr", fmt.Sprintf("%s -- %+v\n", rp.shardID, rp.firehoseWriter.Status()))
```

Comment:
I was unsure if we could invoke the consumer binary with any flags, so I went with env var approach d5467a7c375aa88c8943a5bf282f889f096c88c6